README.md
# Real-time estimation of Omicron and Delta transmission dynamics in England

## Analyses

### Real-time estimation of the time-varying transmission advantage of Omicron in England using S-Gene Target Status as a Proxy

[![Zenodo](https://zenodo.org/badge/435670197.svg)](https://zenodo.org/badge/latestdoi/435670197)

In this work, we use S-gene target failure (SGTF) as a proxy of variant status combined with reported case counts to explore the evidence for changes in transmission advantage over time for the Omicron variant. If present this could indicate the impact of immune escape, bias in SGTF data or differences in the populations within which the variants are circulating. We also report estimates for growth rates by variant and overall, case counts overall and by variant for a 14 day forecast window assuming constant future growth, the date at which Omicron will become dominant in England and in each NHS region, and the estimated cumulative percentage of the population with a reported Omicron case.

Estimates are updated daily as new data is released but the summary of the situation is updated less frequently and so may not match current estimates. All data (both raw data and estimates) are available [here](https://github.com/epiforecasts/omicron-sgtf-forecast/tree/main/data) and the report should be fully reproducible. Reports and data as available at the time of release are available from the [release page](https://github.com/epiforecasts/omicron-sgtf-forecast/releases). See our [news file](https://github.com/epiforecasts/omicron-sgtf-forecast/blob/main/NEWS.md) for details of what updates were made when.

**Real-time report: [as html](https://epiforecasts.io/omicron-sgtf-forecast/summary), [or pdf](https://epiforecasts.io/omicron-sgtf-forecast/summary.pdf)**

### Estimation of the test to test distribution as a proxy for generation interval distribution for the Omicron variant in England

### Background

Early estimates from South Africa indicated that the Omicron COVID-19 variant may be both more transmissible and have greater immune escape than the previously dominant Delta variant. The rapid turnover of the latest epidemic wave in South Africa as well as initial evidence from contact tracing and household infection studies has prompted speculation that the generation time of the Omicron variant may be shorter in comparable settings than the generation time of the Delta variant.

### Methods

We estimated daily growth rates for the Omicron and Delta variants in each UKHSA region from the 23rd of November to the 23rd of December 2021 using surveillance case counts by date of specimen and S-gene target failure status with an autoregressive model that allowed for time-varying differences in the transmission advantage of the Delta variant where the evidence supported this. By assuming a gamma distributed generation distribution we then estimated the generation time distribution and transmission advantage of the Omicron variant that would be required to explain this time varying advantage. We repeated this estimation process using two different prior estimates for the generation time of the Delta variant first based on household transmission and then based on its intrinsic generation time. 

### Results

Visualising our growth rate estimates provided initial evidence for a difference in generation time distributions. Assuming a generation time distribution for Delta with a mean of 2.5-4 days (90% credible interval) and  a standard deviation of 1.9-3 days we estimated a shorter generation time distribution for Omicron with a mean of 1.5-3.2 days and a standard deviation of 1.3-4.6 days. This implied a transmission advantage for Omicron in this setting of 160%-210% compared to Delta. We found similar relative results using an estimate of the intrinsic generation time for Delta though all estimates increased in magnitude due to the longer assumed generation time.

### Conclusions

We found that a reduction in the generation time of Omicron compared to Delta was able to explain the observed variation over time in the transmission advantage of the Omicron variant. However, this analysis cannot rule out the role of other factors such as differences in the populations the variants were mixing in, differences in immune escape between variants or bias due to using the test to test distribution as a proxy for the generation time distribution.

**Paper: [as html](https://epiforecasts.io/omicron-sgtf-forecast/generation-time), [or pdf](https://epiforecasts.io/omicron-sgtf-forecast/generation-time.pdf)**

## Key files and folders

Folder/File | Purpose
---|---
[`writeup`](writeup/) | Summaries and papers as `Rmarkdown` documents.
[`R`](R/) | R functions used in the analysis and for evaluation.
[`data`](data/) | Input data and summarised output generated by steps in the analysis.
[`scripts`](scripts/) | R scripts used to run parts of the analysis.
[`.devcontainer`](.devcontainer/) | Resources for reproducibility using `vscode` and `docker`.

## Dependencies

All dependencies can be installed using the following, 

```r
remotes::install_dev_deps()
```

If `cmdstanr` has not been installed previously then stan may need to be installed. This can be done using the following,

```r
cmdstanr::install_cmdstan()
```

## Access observations and estimates

```r
source(here::here("R", "load-local-data.R"))
get_available_dates()
get_latest_date()
get_local_data(date = <date>)
load_results(date = <date>)
```

## Update estimates

To update the data and estimates run the following. Note this will update estimates regardless of if the data has been updated. 

```bash
bash bin/update-regional-estimates.sh
```

NEWS.md
# 2022-01-07

- Add an analysis estimating the generation time for Omicron that would be required to explain the observed time-varying transmission advantage of Omicron compared to Delta.

# 2021-12-31

- Updated the summary text to reflect post-Christmas changes.
- Updated the time at 95% Omicron plot to include median estimates. This was needed to prevent estimates disappearing when uncertainty was less than or equal to a day.

# 2021-12-29

- Fixed a bug for the centred 7-day moving average where partially complete case data was included for regions. This impacted on the data overview plot.

# 2021-12-28

- Updated data and results archiving to be by date of access rather than by the maximum date available in the data. This has the download of making an update occur every day the pipeline runs regardless of if the data has been updated but the upside of protecting archived data from retrospective updates.

- Updates dating of all archived data and results to be by date of access rather than by maximum available date in the data.

# 2021-12-26

- Automation has been improved so that new data is checked for every two hours. A new model run is only started if not result exist for the current data snapshot or if new SGTF data is detected.

- Updated the plotting of SGTF case and target failure data to be more flexible for other analyses.


# 2021-12-23

- Thanks to [Ritwik Priya](https://twitter.com/ritwik_priya) for spotting an issue with our cumulative incidence as a percentage of the population estimates. Unfortunately, our code accidentally combined all cases types (Omicron, non-Omicron, and combined) when making the calculating leading to estimates roughly 3 times higher than they should have been.
- Update the region order of the cumulative incidence plot to match other regions.

# 2021-12-22

- Initial analysis release.

R/munge-data.R
# Get England SGTF data and clean
library(here)
library(readr)
library(dplyr)
library(lubridate)
library(tidyr)
library(ggplot2)
library(forcats)

add_england_totals <- function(regional) {
  england <- regional %>%
    group_by(date) %>%
    summarise(across(c(total_cases, sgtf_unknown, total_sgt, sgtf, non_sgtf),
                    sum, na.rm = TRUE),
              source = source[1]) %>%
    mutate(region = "England") %>%
    mutate(total_sgt = ifelse(total_sgt == 0, NA, total_sgt)) %>%
    mutate(
      sgtf = ifelse(is.na(total_sgt), NA, sgtf),
      non_sgtf = ifelse(is.na(total_sgt), NA, non_sgtf)
    )

  daily_raw <- bind_rows(england, regional)
  return(daily_raw)
}

tidy_data <- function(regional, region = NULL) {
  # remove "unknown" region (still counted to england total)
  if ("Unknown" %in% unique(regional$region)) {
    regional <- filter(regional, !region == "Unknown")
  }

  subregions <- unique(regional$region)[
    !grepl("^England$", unique(regional$region))
  ]
  regional <- regional %>%
    mutate(region = factor(region, c("England", subregions)))

  # Filter to region, optional
  if (!is.null(region)) {
    region_name <- region
    if (!is.na(region_name)) {
      regional <- filter(regional, region == region_name)}
  }
  return(regional)
}

truncate_cases <- function(regional, days = 0) {
  case_end_date <- max(regional$date) - lubridate::days(days)
  regional <- regional %>%
    mutate(total_cases = ifelse(date > case_end_date, NA, total_cases),
          sgtf_unknown = ifelse(date > case_end_date, NA, sgtf_unknown)) %>%
  return(regional)
}

truncate_sequences <- function(regional, start_date) {
  regional <- regional %>%
    mutate(sgtf = ifelse(date < start_date, NA, sgtf),
           total_sgt = ifelse(date < start_date, NA, total_sgt)) %>%
  return(regional)
}

sgtf_data_to_fv <- function(obs) {
  obs %>%
    transmute(region = region,
              date = date,
              cases = total_cases,
              cases_available = date,
              seq_total = total_sgt,
              seq_voc = sgtf,
              share_voc = sgtf / total_sgt,
              seq_available = date)
}

age_sgtf_data_to_fv <- function(obs) {
  obs %>%
    transmute(region = region,
              age_group = age_group,
              date = date,
              cases = total_cases,
              cases_available = date,
              seq_total = total_sgt,
              seq_voc = sgtf,
              share_voc = sgtf / total_sgt,
              seq_available = date)
}

bias_data_to_fv <- function(obs) {
  obs %>%
    transmute(region = region,
              date = date,
              cases = total_cases,
              seq_total = total_cases,
              seq_voc = total_sgt,
              share_voc = total_sgt / total_cases,
              cases_available = date,
              seq_available = date) %>%
    filter(!is.na(seq_total)) %>%
    mutate(seq_total = ifelse(is.na(seq_voc), NA, seq_total)) %>%
    filter(!(is.na(cases) & is.na(seq_voc))) %>%
    filter(!is.na(seq_total))
}

age_seq_data_to_fv <- function(obs) {
  obs %>%
    transmute(region = region,
              age_group = age_group,
              date = date_specimen,
              cases = total,
              cases_available = date,
              seq_total = total_seq,
              seq_voc = omicron,
              share_voc = omicron / total_seq,
              seq_available = date)
}

filter_seq_threshold <- function(obs, frac_thres = 0.01, n_thres = 5) {
  obs %>%
    mutate(great_thres = share_voc >= frac_thres & seq_voc >= n_thres,
           start_date = ifelse(great_thres, date, NA),
           start_date = min(start_date, na.rm = TRUE)) %>%
    mutate(across(.cols = c("seq_total", "seq_voc", "share_voc"),
           ~ ifelse(date < start_date, NA, .))
    ) %>%
    select(-start_date, -great_thres)
}

cumulative_percentage <- function(cases, pop) {
  cases_pop <- cases %>%
    left_join(pop, by = c("region")) %>%
    group_by(region, type) %>%
    arrange(date) %>%
    mutate(c_median = cumsum(median) / population,
           c_q5 = cumsum(q5) / population,
           c_q95 = cumsum(q95) / population,
           c_q20 = cumsum(q20) / population,
           c_q80 = cumsum(q80) / population) %>%
    ungroup()
  return(cases_pop)
}

summarise_by_15_year_age_group <- function(dt) {
  dt %>%
    dplyr::mutate(
      age_group = forcats::fct_collapse(age_group,
        `0-14` = c("0-4", "5-9", "10-14"),
        `15-29` = c("15-19", "20-24", "25-29"),
        `30-44` = c("30-34", "35-39", "40-44"),
        `45-59` = c("45-49", "50-54", "55-59"),
        `60-74` = c("60-64", "65-69", "70-74"),
        `75+` = c("75-79", "80-84", "85-89", "90+")
    )
  ) %>%
    group_by(age_group, region, date) %>%
    summarise(across(where(is.numeric), sum, na.rm = TRUE)) %>%
    ungroup()
}